src/tools/tidy/src/watcher.rs

```diff
@@ -0,0 +1,193 @@
+//! Checks that text between tags unchanged, emitting warning otherwise,
+//! allowing asserting that code in different places over codebase is in sync.
+//!
+//! This works via hashing text between tags and saving hash in tidy.
+//!
+//! Usage:
+//!
+//! some.rs:
+//! // tidy-ticket-foo
+//! const FOO: usize = 42;
+//! // tidy-ticket-foo
+//!
+//! some.sh:
+//! # tidy-ticket-foo
+//! export FOO=42
+//! # tidy-ticket-foo
+use md5::{Digest, Md5};
+use std::fs;
+use std::path::Path;
+
+#[cfg(test)]
+mod tests;
+
+/// Return hash for source text between 2 tag occurrence,
+/// ignoring lines where tag written
+///
+/// Expecting:
+/// tag is not multiline
+/// source always have at least 2 occurrence of tag (>2 ignored)
+fn span_hash(source: &str, tag: &str, bad: &mut bool) -> Result<String, ()> {
+    let start_idx = match source.find(tag) {
+        Some(idx) => idx,
+        None => return Err(tidy_error!(bad, "tag {} should exist in provided text", tag)),
+    };
+    let end_idx = {
+        let end = match source[start_idx + tag.len()..].find(tag) {
+            // index from source start
+            Some(idx) => start_idx + tag.len() + idx,
+            None => return Err(tidy_error!(bad, "tag end {} should exist in provided text", tag)),
+        };
+        // second line with tag can contain some other text before tag, ignore it
+        // by finding position of previous line ending
+        //
+        // FIXME: what if line ending is \r\n? In that case \r will be hashed too
+        let offset = source[start_idx..end].rfind('\n').unwrap();
+        start_idx + offset
+    };
+
+    let mut hasher = Md5::new();
+
+    source[start_idx..end_idx]
+        .lines()
+        // skip first line with tag
+        .skip(1)
+        // hash next lines, ignoring end trailing whitespaces
+        .for_each(|line| {
+            let trimmed = line.trim_end();
+            hasher.update(trimmed);
+        });
+    Ok(format!("{:x}", hasher.finalize()))
+}
+
+fn check_entry(entry: &ListEntry<'_>, group_idx: usize, bad: &mut bool, root_path: &Path) {
+    let file = fs::read_to_string(root_path.join(Path::new(entry.0)))
+        .unwrap_or_else(|e| panic!("{:?}, path: {}", e, entry.0));
+    let actual_hash = span_hash(&file, entry.2, bad).unwrap();
+    if actual_hash != entry.1 {
+        // Write tidy error description for wather only once.
+        // Will not work if there was previous errors of other types.
+        if *bad == false {
+            tidy_error!(
+                bad,
+                "Mismatched hashes for tidy watcher found.\n\
+                Check src/tools/tidy/src/watcher.rs, find tag/hash in TIDY_WATCH_LIST list \
+                and verify that sources for provided group of tags in sync. If they in sync, update hash."
```

Comment:
This needs a way to bless it automatically using `--bless`, otherwise it could be very annoying with big refactors.

Reply:
Bless will be nice, but there currently no usage experience; maybe this need to be changed only few times in month, who knows.